src/main/scala/io/iohk/ethereum/db/cache/LruCache.scala

```diff
@@ -0,0 +1,48 @@
+package io.iohk.ethereum.db.cache
+
+import java.util.concurrent.TimeUnit
+import java.util.concurrent.atomic.AtomicLong
+import io.iohk.ethereum.utils.Config.NodeCacheConfig
+import com.google.common.cache
+import com.google.common.cache.{CacheBuilder, RemovalNotification}
+import scala.concurrent.duration.FiniteDuration
+
+class LruCache[K <: AnyRef, V <: AnyRef](config: NodeCacheConfig, notificationHandler: Option[RemovalNotification[K, V] => Unit] = None)
+  extends Cache[K, V] {
+
+  private val lastClear = new AtomicLong(System.nanoTime())
+
+  private val lruCache : cache.Cache[K, V] =
+    CacheBuilder.newBuilder()
+      .maximumSize(config.maxSize)
+      .removalListener(
+        (notification: RemovalNotification[K, V]) => if (notification.wasEvicted()) {
+          notificationHandler.foreach(handler => handler(notification))
+        }
+      )
+      .build()
+
+  override def clear(): Unit = {
+    lastClear.getAndSet(System.nanoTime())
+    lruCache.invalidateAll()
+  }
+
+  override def getValues: Seq[(K, V)] = {
+    import scala.collection.JavaConverters._
+    lruCache.asMap().asScala.toSeq
```

Comment:
i am not sure what you mean.  I do not want to evict any entries here, just get all available in cache, there is separate method for caller to invalidate all entries.

Reply:
Yes, but this method doesn't guarantee that all entries are valid. Calling "invalidateAll" will do the job but I am not sure if we want to do that from caller.

Reply:
` doesn't guarantee that all entries are valid`  you mean that some other thread could manipulate the cache during this operation, or between the calls to `getValues` and `clear` ? 
Actually in this approach there is this silent assumption that there will always be at most one writer thread, and the same thread which is doing the updates will later do pruning. As for now this is preserved, becouse we are processing each whole block in same thread. But maybe some comment is needed here to explain that , wdyt ?

Reply:
For example, if you will use time eviction the thread responsible for maintaining the map could not clear the evicted entries from map yet. In this case it is probably not a problem. I am not sure how it behaves when you are using the max size limit.

Reply:
actually from google cache docs `Caches built with CacheBuilder do not perform cleanup and evict values "automatically," or instantly after a value expires, or anything of the sort. Instead, it performs small amounts of maintenance during write operations, or during occasional read operations if writes are rare.`  so there is no maintance thread unlees explictly created.
Also iterating methods like `asMap` do not trigger eviction.

Reply:
Ok, but it still doesn't solve a problem. There is a possibility that you return evicted entries (because of size limit) before mentioned maintenance takes place. Does it?

Reply:
Hmm i think i do not follow, current execution of block (which happens in single thread) looks something like that:
`1.updates and additions to cache due to tx execution -> 
2.pruning (deleting from cache unrefereneced elements) -> 3.  every one hour getVals from cache and persist them in storage then clear cache`
All eviction can happen during phase `1` and maybe `2`, so during phase `3` all unnecessery elements were already evicted and we wont return them.

Reply:
I think it will be work in most cases and it could be an academic discussion :P My point is that invalidate != cleanUp so if you call `getValues` before maintenance happens you get already invalid entries.